README.md
# www-salesforce-nb [![Build Status](https://travis-ci.org/genio/www-salesforce-nb.svg?branch=master)](https://travis-ci.org/genio/www-salesforce-nb)

A non-blocking [Salesforce API](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_what_is_rest_api.htm) client using [Mojolicious](http://mojolicio.us). It may one day replace [WWW::Salesforce](http://metacpan.org/pod/WWW::Salesforce).

It is EXTREMELY experimental at this point.  Use it at your own risk.  You've been warned.

## Table of Contents

* [Synopsis](#synopsis)
* [Description](#description)
* [Events](#events)
* [Attributes](#attributes)
	* [consumer\_key](#consumer_key)
	* [consumer\_secret](#consumer_secret)
	* [login\_type](#login_type)
		* [oauth2\_up](#oauth2_up)
		* [soap](#soap)
	* [login\_url](#login_url)
	* [pass\_token](#pass_token)
	* [password](#password)
	* [ua](#ua)
	* [username](#username)
	* [version](#version)
* [Methods](#methods)
	* [create](#create)
	* [describe](#describe)
	* [limits](#limits)
	* [login](#login)
	* [logout](#logout)
	* [query](#query)
	* [retrieve](#retrieve)
	* [search](#search)
* [Error Handling](#error-handling)
* [Author](#author)
* [Bugs](#bugs)

## SYNOPSIS

```perl
# via soap
my $sf_soap = WWW::Salesforce->new(
	login_type => 'soap',
	login_url => Mojo::URL->new('https://login.salesforce.com'),
	version => '34.0',
	username => 'foo@bar.com',
	password => 'mypassword',
	pass_token => 'mypasswordtoken123214123521345',
);
# via OAuth2 username and password
my $sf_oauth2 = WWW::Salesforce->new(
	login_type => 'oauth2_up', # this is the default
	login_url => Mojo::URL->new('https://login.salesforce.com'),
	version => '34.0',
	consumer_key => 'alksdlkj3hasdg;jlaksghajdhgaghasdg.asdgfasodihgaopih.asdf',
	consumer_secret => 'asdfasdjkfh234123513245',
	username => 'foo@bar.com',
	password => 'mypassword',
	pass_token => 'mypasswordtoken123214123521345',
);

# blocking method
# calling login() will happen automatically.
try {
	my $res_soap = $sf_soap->query('Select Id, Name, Phone from Account');
	say "found ", scalar(@{$res_soap}), " results via SOAP then RESTful API.";
	my $res_oauth = $sf_oauth2->query('Select Id, Name, Phone from Account');
	say "found ", scalar(@{$res_oauth}), " results via OAuth2 then RESTful API.";
}
catch {
	die "Couldn't query the service: $_";
};

# non-blocking method
# calling login() will happen automatically
Mojo::IOLoop->delay(
	sub {
		my $delay = shift;
		$sf_soap->query('select Id from Account', $delay->begin);
		$sf_oauth2->query('select Id from Account', $delay->begin);
	},
	sub {
		my ($delay, $err,$soap,$err2,$oauth) = @_;
		Carp::croak( $err ) if $err; # make it fatal
		Carp::croak( $err2 ) if $err2; # make it fatal
		say scalar(@$soap), " from soap";
		say scalar(@$oauth), " from oauth2";
	},
)->catch(sub {say "uh oh, ",pop;})->wait;
```

## DESCRIPTION

[WWW::Salesforce](https://github.com/genio/www-salesforce-nb/) allows us to connect to [Salesforce](http://www.salesforce.com/)'s service to access our data using their [RESTful API](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/).

Creation of a new [WWW::Salesforce](https://github.com/genio/www-salesforce-nb/) instance will not actually hit the server.  The first communication with the [Salesforce](http://www.salesforce.com/) API occurs when you specifically call the ```login``` method or when you make another call.

All API calls using this library will first make sure you are properly logged in.

* ```oauth2_up``` OAuth2 using the username and password (up) method:
	* [Session ID Authorization](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/quickstart_oauth.htm)
	* [Salesforce Username-Password OAuth Authentication Flow](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_understanding_username_password_oauth_flow.htm)
* ```soap``` Alternately, you can use the soap login method:
	* [Salesforce SOAP-based username and password login flow](https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_calls_login.htm)

## ATTRIBUTES

[WWW::Salesforce](https://github.com/genio/www-salesforce-nb/) makes the following attributes available.

### consumer\_key

```perl
my $key = $sf->consumer_key;
$sf = $sf->consumer_key( 'alksdlksdf' ); # method-chaining
```

The Consumer Key (also referred to as the client\_id in the Saleforce documentation) is part of your [Connected App](http://www.salesforce.com/us/developer/docs/api_rest/Content/intro_defining_remote_access_applications.htm).  It is a required field to be able to login.

Note, this attribute is only used to generate the access token during [login](#login).  You may want to [logout](#logout) before changing this setting.

### consumer\_secret

```perl
my $secret = $sf->consumer_secret;
$sf = $sf->consumer_secret( 'asdfas123513245' ); # method-chaining
```

The Consumer Secret (also referred to as the client\_secret in the Saleforce documentation) is part of your [Connected App](http://www.salesforce.com/us/developer/docs/api_rest/Content/intro_defining_remote_access_applications.htm).  It is a required field to be able to login.

Note, this attribute is only used to generate the access token during [login](#login).  You may want to [logout](#logout) before changing this setting.

### login\_type

```perl
my $type = $sf->login_type;
$sf = $sf->login_type( 'oauth2_up' ); # method-chaining
```

This is what will determine our login method of choice. No matter which login method you choose, we're going to communicate to the Salesforce services using an ```Authorization: Bearer token``` header. The login method just dictates how we will request that token from Salesforce.  Different methods of login require slightly different sets of data in order for the login to take place.

You may want to [logout](#logout) before changing this setting.

Available types are:

#### oauth2_up

This login type is the default.  It will require your [consumer\_key](#consumer_key), [consumer\_secret](#consumer_secret), [username](#username), [password](#password), [pass\_token](#pass_token) and [login\_url](#login_url).  This method will go through the [Salesforce Username-Password OAuth Authentication Flow](http://www.salesforce.com/us/developer/docs/api_rest/Content/intro_understanding_username_password_oauth_flow.htm).


#### soap

This method will only require your [username](#username), [password](#password), [pass\_token](#pass_token) and [login\_url](#login_url). It will go through the [Salesforce SOAP-based username and password login flow](https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_calls_login.htm).

### login\_url

```perl
my $host = $sf->login_url;
$sf = $sf->login_url( Mojo::URL->new('https://test.salesforce.com') ); # method-chaining
```

This is the base host of the API we're using.  This allows you to use any of your sandbox or live data areas easily. You may want to [logout](#logout) before changing this setting.

### pass\_token

```perl
my $token = $sf->pass_token;
$sf = $sf->pass_token( 'mypasswordtoken145' ); # method-chaining
```

The password token is a Salesforce-generated token to go along with your password.  It is appended to the end of your password and used only during [login](#login) authentication.

Note, this attribute is only used to generate the access token during [login](#login).  You may want to [logout](#logout) before changing this setting.

### password

```perl
my $password = $sf->password;
$sf = $sf->password( 'mypassword' ); # method-chaining
```

The password is the password you set for your user account in Salesforce.

Note, this attribute is only used to generate the access token during [login](#login).  You may want to [logout](#logout) before changing this setting.

### ua

```perl
my $ua = $sf->ua;
```

The [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) is the user agent we use to communicate with the Salesforce services.  For ```proxy``` and other needs, see the [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) documentation.

### username

```perl
my $username = $sf->username;
$sf = $sf->username( 'foo@bar.com' ); # method-chaining
```

The username is the email address you set for your user account in Salesforce.
Note, this attribute is only used to generate the access token during [login](#login).  You may want to [logout](#logout) before changing this setting.

### version

```perl
my $version = $sf->version;
$sf = $sf->version( '34.0' ); # method-chaining
```

Tell us what API version you'd like to use.  Leave off the ```v``` from the version number.


## METHODS

[WWW::Salesforce](https://github.com/genio/www-salesforce-nb/) makes the following methods available.

### create

```perl
# blocking
try {
	my $res = $sf->create('Account',{fieldName=>'value'});
	if ( $res->{success} ) { # even if the tx succeeds, check the response!
		say "Newly entered Account goes by the id: ",$res->{id};
	}
	else {
		die Dumper $res->{errors};
	}
} catch {
	die "Errors: $_";
};

# non-blocking
$sf->create('Account',{fieldName=>'value'}, sub {
	my ($sf, $err, $res) = @_;
	die "Got an error trying to create the Account: $err" if $err;
	if ( $res->{success} ) { # even if the tx succeeds, check the response!
		say "Newly entered Account goes by the id: ",$res->{id};
	}
	else {
		die Dumper $res->{errors};
	}
});
```

This method calls the Salesforce [Create method](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/dome_sobject_create.htm).
On a successful transaction, a JSON response is returned with three fields (```id```, ```success```, and ```errors```).  You should check that response to see if your creation attempt actually succeeded.

### describe

```perl
# blocking
try {
	my $res = $sf->describe('Account');
	say Dumper $res; #all the info about the Account SObject
} catch {
	die "Errors: $_";
};

# non-blocking
$sf->describe('Account', sub {
	my ($sf, $err, $res) = @_;
	die "Got an error trying to describe the Account: $err" if $err;
	say Dumper $res; #all the info about the Account SObject
});
```

This method calls the Salesforce [Describe method](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_sobject_describe.htm).
On a successful transaction, a JSON response is returned with data full of useful information about the SObject.

### limits

```perl
# blocking
try {
	my $results = $sf->limits();
	say Dumper $results;
} catch {
	die "Errors: $_";
};

# non-blocking
$sf->limits(sub {
	my ($sf, $err, $results) = @_;
	say Dumper $results;
});
```

This method calls the Salesforce [Limits method](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_limits.htm).

### login

```perl
## blocking
$sf = $sf->login(); # allows for method-chaining

## non-blocking
$sf->login( sub {
	my ($sf, $token) = @_;
	say "Our auth token is: $token";
});
```

This method will go through the [Salesforce Username-Password OAuth Authentication Flow](http://www.salesforce.com/us/developer/docs/api_rest/Content/intro_understanding_username_password_oauth_flow.htm) process if it needs to.
Calling this method on your own is not necessary as any API call will call ```login``` if necessary.  This could be helpful if you're changing ```api_host```s on your instance.
This method will update your ```access_token``` on a successful login.

### logout

```perl
$sf = $sf->logout(); # allows for method-chaining
```

This method does not actually make any call to [Salesforce](http://www.salesforce.com).
It only removes knowledge of your access token so that you can login again on your next API call.

### query

```perl
## blocking
my $results = $sf->query('Select Id, Name, Phone from Account');
say Dumper $results;

## non-blocking
$sf->query('select Id, Name, Phone from Account', sub {
	my ($sf, $results) = @_;
	say Dumper $results;
});
```

This method calls the Salesforce [Query method](http://www.salesforce.com/us/developer/docs/api_rest/Content/resources_query.htm).  It will keep grabbing and adding the records to your resultant array reference until there are no more records available to your query.

Your query string must be in the form of an [SOQL String](https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_select.htm).

### retrieve

```perl
# blocking
try {
	my $results = $sf->retrieve('Account','01231ABCDFQ2100002', [qw(Optional FieldName List Here)]);
	# can also call as ->retrieve('Account','01231ABCDFQ2100002')
	say Dumper $results;
} catch {
	die "Errors: $_";
};

# non-blocking
$sf->retrieve('Account','01231ABCDFQ2100002', [qw(Optional FieldName List Here)], sub {
	# can also call as ->retrieve('Account','01231ABCDFQ2100002', sub {...});
	my ($sf, $err, $results) = @_;
	say Dumper $results;
});
```

This method calls the Salesforce [Retrieve method](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/dome_get_field_values.htm).

### search

```perl
# blocking
try {
	my $results = $sf->search('FIND{genio*}');
	say Dumper $results;
} catch {
	die "Errors: $_";
};

# non-blocking
$sf->search('FIND{genio*}', sub {
	my ($sf, $err, $results) = @_;
	say Dumper $results;
});
```

This method calls the Salesforce [Search method](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_search.htm).
Your search query must be in the form of an [SOSL String](https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_sosl_syntax.htm).

## ERROR HANDLING

All blocking method calls will ```die``` on error and thus you should use [Try::Tiny](https://metacpan.org/pod/Try::Tiny) a lot.

```perl
# blocking call
use Try::Tiny qw(try catch);
try {
	my $res = $sf->do_something();
} catch {
	die "uh oh: $_";
};
```

All non-blocking methods will return an error string to the callback if there is one:

```perl
# non-blocking call
$sf->do_something(sub {
	my ( $instance, $error_string, $results ) = @_;
	die "uh oh: $error_string" if $error_string;
});
```

## AUTHOR

Chase Whitener -- cwhitener@gmail.com

## BUGS

Please report any bugs or feature requests on GitHub [https://github.com/genio/www-salesforce-nb/issues](https://github.com/genio/www-salesforce-nb/issues).
We appreciate any and all criticism, bug reports, enhancements, or fixes.


Changes
Revision history for Perl extension WWW::Salesforce.

0.011
	- Simplify the mock servers a bit
	- Fix the non-blocking tests to ensure they all run

0.010
	- Pull the ::Connector methods back into WWW::Salesforce directly
	- Change from a Moo object to a Mojo::Base object
	- Got coverage tests up a slight bit

0.009
	- All NB calls use Mojo::IOLoop->delay() now.
	- All NB early errors use next_tick
	- All coverage upped above 97%

0.008
	- moved the test mock server into its own file
	- made the Connector a role
	- removed a superfluous header for SOAP calls
	- added tests for the blocking create method

0.007
	- added create method
	- added describe method
	- added limits method
	- added retrieve method
	- added search method
	- fixed some blocking error handling
	- fixed some docs

0.006
	- split logins out to the Connector class.
	- allows for soap and oauth2 based logins

0.005
	- die on error in blocking code
	- change non-blocking to return $self, $err, $data on error
	- stop emitting error events
	- cleaned up some

0.004
	- Stopped subclassing as it was not necessary
	- delegated certain methods/attributes instead

0.003
	- made it a Mojo::UserAgent instead of Mojo::EventEmitter

0.002
	- created logout()
	- updated lots of documentation

0.001
	- created login() for both blocking and non-blocking
	- created query() for both blocking and non-blocking
	- created _require_login() to keep our connection alive for long periods


t/lib/mock.pl
use Mojolicious::Lite;
use Data::Dumper;
BEGIN {
	$ENV{MOJO_NO_SOCKS} = $ENV{MOJO_NO_TLS} = 1;
	$ENV{MOJO_REACTOR} = 'Mojo::Reactor::Poll';
}

# Silence
app->log->level('fatal');
get '/' => {text => 'works!'};
get '/services/data/v33.0/sobjects' => sub {
	my $c = shift;
	return $c->render(json=>{
		encoding => "UTF-8",
		maxBatchSize => 200,
		sobjects => [
			{
				activateable => 'false',
				createable => 'true',
				custom => 'false',
				customSetting => 'false',
				deletable => 'true',
				deprecatedAndHidden => 'false',
				feedEnabled => 'true',
				keyPrefix => "001",
				label => "Account",
				labelPlural => "Accounts",
				layoutable => 'true',
				mergeable => 'true',
				name => "Account",
				queryable => 'true',
				replicateable => 'true',
				retrieveable => 'true',
				searchable => 'true',
				triggerable => 'true',
				undeletable => 'true',
				updateable => 'true',
				urls => {
					compactLayouts => "/services/data/v34.0/sobjects/Account/describe/compactLayouts",
					rowTemplate => "/services/data/v34.0/sobjects/Account/{ID}",
					approvalLayouts => "/services/data/v34.0/sobjects/Account/describe/approvalLayouts",
					listviews => "/services/data/v34.0/sobjects/Account/listviews",
					describe => "/services/data/v34.0/sobjects/Account/describe",
					quickActions => "/services/data/v34.0/sobjects/Account/quickActions",
					layouts => "/services/data/v34.0/sobjects/Account/describe/layouts",
					sobject => "/services/data/v34.0/sobjects/Account",
				},
			},
		],
	});
};
post '/services/data/v33.0/sobjects/:type' => sub {
	my $c = shift;
	my $type = $c->stash('type');
	my $params = $c->req->json;
	return $c->render(json=>{success=>'false',id=>undef,errors=>['bad object']},status=>500) unless $type;
	return $c->render(json=>{success=>'false',id=>undef,errors=>['no params']},status=>500) unless $params && ref($params) eq 'HASH';
	return $c->render(json=>{success=>'false',id=>undef,errors=>['bad object']}) unless $type eq 'Account';
	return $c->render(json=>{success=>'true',id=>'01t500000016RuaAAE',errors=>[]});
};
get '/services/data/v33.0/query' => sub {
	my $c = shift;
	my $query = $c->param('q');
	if ( $query eq 'select Id,IsActive,Name from Product2' ) {
		return $c->render(json => {done=>0,nextRecordsUrl=>'/services/data/v33.0/query/test123',records=>[
			{
				Id=>'01t500000016RuaAAE',
				IsActive=>1,
				Name=>'Test Name',
				attributes=>{
					type=>'Product2',
					url=>'/services/data/v33.0/sobjects/Product2/01t500000016RuaAAE',
				},
			},
		],});
	}
	$c->render(json=>[{errorCode=>'foo',message=>'what?!?'}], status=>401);
};
get '/services/data/v33.0/query/test123' => sub {
	my $c = shift;
	return $c->render(json => {done=>1,records=>[
		{
			Id=>'01t500000016RuaAAF',
			IsActive=>1,
			Name=>'Test Name 2',
			attributes=>{
				type=>'Product2',
				url=>'/services/data/v33.0/sobjects/Product2/01t500000016RuaAAF',
			},
		},
	],});
};
post '/services/Soap/u/33.0/' => sub {
	my $c = shift;
	my $username = '';
	my $password = '';
	my $input = '';
	$input = $c->req->dom() if $c->req && $c->req->content && $c->req->dom;
	my $eof = q(<?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"><soapenv:Body><soapenv:Fault><faultcode>soapenv:Client</faultcode><faultstring>Premature end of file.</faultstring></soapenv:Fault></soapenv:Body></soapenv:Envelope>);
	return $c->render(data=>$eof, format => 'xml', status=>500) unless $input;
	$username = $input->at('urn\:username')->text() if $input->at('urn\:username');
	$password = $input->at('urn\:password')->text() if $input->at('urn\:password');
	#return actual error messages
	my $error=q(<?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:sf="urn:fault.partner.soap.sforce.com" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><soapenv:Body><soapenv:Fault><faultcode>sf:INVALID_LOGIN</faultcode><faultstring>INVALID_LOGIN: Invalid username, password, security token; or user locked out.</faultstring><detail><sf:LoginFault xsi:type="sf:LoginFault"><sf:exceptionCode>INVALID_LOGIN</sf:exceptionCode><sf:exceptionMessage>Invalid username, password, security token; or user locked out.</sf:exceptionMessage></sf:LoginFault></detail></soapenv:Fault></soapenv:Body></soapenv:Envelope>);
	return $c->render(data=>$error, format => 'xml', status=>500) unless $username eq 'test';
	return $c->render(data=>$error, format => 'xml', status=>500) unless $password eq 'testtoke';
	# return the successful response
	my $success=qq(<?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns="urn:partner.soap.sforce.com" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><soapenv:Body><loginResponse><result><metadataServerUrl>/</metadataServerUrl><passwordExpired>false</passwordExpired><sandbox>false</sandbox><serverUrl>/</serverUrl><sessionId>123455663452abacbabababababababanenenenene</sessionId><userId>00e30658AA0de34AA2</userId><userInfo><accessibilityMode>false</accessibilityMode><currencySymbol>\$</currencySymbol><orgAttachmentFileSizeLimit>5242880</orgAttachmentFileSizeLimit><orgDefaultCurrencyIsoCode>USD</orgDefaultCurrencyIsoCode><orgDisallowHtmlAttachments>false</orgDisallowHtmlAttachments><orgHasPersonAccounts>false</orgHasPersonAccounts><organizationId>00e30658AA0de34AAX</organizationId><organizationMultiCurrency>false</organizationMultiCurrency><organizationName>Test Company</organizationName><profileId>00e30658AA0de34AAA</profileId><roleId>00e30658AA0de34AA1</roleId><sessionSecondsValid>14400</sessionSecondsValid><userDefaultCurrencyIsoCode xsi:nil="true"/><userEmail>test\@tester.com</userEmail><userFullName>Test User</userFullName><userId>00e30658AA0de34AA2</userId><userLanguage>en_US</userLanguage><userLocale>en_US</userLocale><userName>$username</userName><userTimeZone>America/New_York</userTimeZone><userType>Standard</userType><userUiSkin>Theme3</userUiSkin></userInfo></result></loginResponse></soapenv:Body></soapenv:Envelope>);
	return $c->render(data => $success, format => 'xml');
};
post '/services/oauth2/revoke' => sub {
	my $c = shift;
	my $token = $c->param('token');
	return $c->render(json=>[{error_description=>"invalid token: $token",error=>"unsupported_token_type"}], status=>400) unless $token eq '123455663452abacbabababababababanenenenene';
	return $c->render(json=>[{success=>'true'}]);
};
post '/services/oauth2/token' => sub {
	my $c = shift;
	my $grant_type = $c->param('grant_type') || '';
	my $client_id = $c->param('client_id') || '';
	my $client_secret = $c->param('client_secret') || '';
	my $username = $c->param('username') || '';
	my $password = $c->param('password') || '';
	#return actual error messages
	return $c->render(json=>[{error_description=>"grant type not supported",error=>"unsupported_grant_type"}], status=>400) unless $grant_type eq 'password';
	return $c->render(json=>[{error_description=>'Invalid client credentials',error=>'invalid_client'}], status=>400) unless $client_id eq 'test_id';
	return $c->render(json=>[{error_description=>'Invalid client credentials',error=>'invalid_client'}], status=>400) unless $client_secret eq 'test_secret';
	return $c->render(json=>[{error_description=>'authentication failure',error=>'invalid_grant'}], status=>400) unless $username eq 'test';
	return $c->render(json=>[{error_description=>'authentication failure',error=>"invalid_grant"}], status=>400) unless $password eq 'testtoke';
	# return the successful response
	return $c->render(json => {
		id => "/id/00D3012300VnRVAU/0015004310HWV5ZAQ",
		token_type =>"Bearer",
		signature => "CtSomeSignature3421351345141LKJFSDLK8723451nhx8=",
		instance_url => Mojo::URL->new('/'),
		issued_at => time()*1000,
		access_token => '123455663452abacbabababababababanenenenene'
	});
};
app->start;


Makefile.PL
use 5.010001;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
	NAME => 'WWW::Salesforce',
	VERSION_FROM => 'lib/WWW/Salesforce.pm',
	ABSTRACT => 'A non-blocking Perl Salesforce client',
	AUTHOR => q{Chase Whitener <cwhitener@gmail.com>},
	LICENSE => 'artistic_2',
	META_MERGE => {
		dynamic_config => 0,
		'meta-spec' => {version=>2},
		no_index => {directory => ['t']},
		prereqs => {runtime => {requires => {perl => '5.010001'}}},
		resources => {
			bugtracker => {web=>'https://github.com/genio/www-salesforce-nb/issues'},
			homepage => 'https://github.com/genio/www-salesforce-nb',
			license => ['http://www.opensource.org/licenses/artistic-license-2.0'],
			repository => {
				type => 'git',
				url => 'https://github.com/genio/www-salesforce-nb.git',
				web => 'https://github.com/genio/www-salesforce-nb',
			},
			X_IRC => 'irc://irc.perl.org/#perl',
		},
	},
	BUILD_REQUIRES => {
		'Test::More' => 0,
		'Test::Pod' => 0,
		'Test::Pod::Coverage' => 0,
		'Try::Tiny' => 0,
	},
	PREREQ_PM => {
		'IO::Socket::IP' => 0.20,
		'IO::Socket::Socks' => 0.64,
		'IO::Socket::SSL' => 1.84,
		'Mojolicious' => 5.24,
		'Moo' => 0.0,
		'namespace::clean' => 0.0,
		'POSIX' => 0.0,
		'strictures' => 2.0,
		'utf8' => 0.0,
	},
	test => {TESTS => 't/*.t'},
);


lib/WWW/Salesforce/Connector.pm
package WWW::Salesforce::Connector;

use strictures 2;
use Mojo::URL ();
use Mojo::Util qw(xml_escape);
use Scalar::Util ();
use WWW::Salesforce::SOAP;

use Moo::Role;
use 5.010;

# attempt a login to Salesforce to obtain a token
sub login {
	my $cb = ($_[-1] && ref($_[-1]) eq 'CODE')? pop: undef;
	my ($self,$username,$password) = @_;
	unless ( $self->_login_required ) {
		Mojo::IOLoop->next_tick(sub { $self->$cb(undef, $self->_access_token) }) if $cb;
		return $self;
	}
	my $type = $self->login_type();
	return $self->_login_soap($cb) if $type eq 'soap';
	return $self->_login_oauth2_up($cb)
}
# log out of salesforce and invalidate the token we're using
sub logout {
	my $cb = ($_[-1] && ref($_[-1]) eq 'CODE')? pop: undef;
	my $self = shift;
	if ( $self->_login_required ) {
		# we don't need to logout
		Mojo::IOLoop->next_tick(sub { $self->$cb(undef,1) }) if $cb;
		return $self;
	}
	my $headers = $self->_headers();
	$headers->{content_type} = 'application/x-www-form-urlencoded';
	my $url = Mojo::URL->new($self->_instance_url)->path("/services/oauth2/revoke");
	#blocking request
	unless ( $cb ) {
		my $tx = $self->ua->post($url, $self->_headers(), form =>{token=>$self->_access_token});
		die $self->_error($tx) unless $tx->success;
		$self->_instance_url(undef);
		$self->_access_token(undef);
		$self->_access_time(0);
		$self->ua->cookie_jar->empty();
		return $self;
	}
	# non-blocking request
	Mojo::IOLoop->delay(
		sub { $self->ua->post($url, $self->_headers(), form =>{token=>$self->_access_token}, shift->begin); },
		sub {
			my ($delay, $tx) = @_;
			return $self->$cb($self->_error($tx),undef) unless $tx->success;
			$self->_instance_url(undef);
			$self->_access_token(undef);
			$self->_access_time(0);
			$self->ua->cookie_jar->empty();
			return $self->$cb(undef, 1);
		}
	)->catch(sub {
		return $self->$cb(pop);
	})->wait;
	return $self;
}

sub _login_oauth2_up {
	my ( $self, $cb ) = @_;
	my $url = Mojo::URL->new($self->login_url)->path("/services/oauth2/token");
	my $form = {
		grant_type => 'password',
		client_id => $self->consumer_key,
		client_secret => $self->consumer_secret,
		username => $self->username,
		password => $self->password . $self->pass_token,
	};
	# blocking request
	unless ($cb) {
		my $tx = $self->ua->post($url, $self->_headers(), form => $form);
		die $self->_error($tx) unless $tx->success;
		my $data = $tx->res->json;
		$self->_instance_url(Mojo::URL->new($data->{instance_url}));
		$self->_access_token($data->{access_token});
		$self->_access_time($data->{issued_at}/1000); #convert milliseconds to seconds
		return $self;
	}

	# non-blocking request
	Mojo::IOLoop->delay(
		sub { $self->ua->post($url, $self->_headers(), form =>$form, shift->begin); },
		sub {
			my ($delay, $tx) = @_;
			return $self->$cb($self->_error($tx),undef) unless $tx->success;
			my $data = $tx->res->json;
			$self->_instance_url(Mojo::URL->new($data->{instance_url}));
			$self->_access_token($data->{access_token});
			$self->_access_time($data->{issued_at}/1000); #convert milliseconds to seconds
			return $self->$cb(undef, $self->_access_token);
		}
	)->catch(sub {
		return $self->$cb(pop);
	})->wait;
	return $self;
}

# returns true (1) if login required, else undef
sub _login_required {
	my $self = shift;
	if( $self->_access_token && $self->_instance_url ) {
		my $time = $self->_access_time;
		return undef if ( int((time() - $time)/60) < 30 );
	}
	$self->ua->cookie_jar->empty();
	return 1;
}

sub _login_soap {
	my ( $self, $cb ) = @_;
	my $url = Mojo::URL->new($self->login_url)->path($self->_path('soap'));
	my $envelope = WWW::Salesforce::SOAP::envelope_login($self->username, $self->password, $self->pass_token)->to_string;

	unless ( $cb ) {
		my $tx = $self->ua->post($url, $self->_headers('soap'), $envelope);
		die $self->_error($tx) unless $tx->success;
		my $data = WWW::Salesforce::SOAP::response_login($tx->res->dom);
		$self->_instance_url(Mojo::URL->new($data->{serverUrl}));
		$self->_access_token($data->{sessionId});
		$self->_access_time(time);
		return $self;
	}
	# non-blocking request
	Mojo::IOLoop->delay(
		sub { $self->ua->post($url, $self->_headers(), $envelope, shift->begin); },
		sub {
			my ($delay, $tx) = @_;
			return $self->$cb($self->_error($tx),undef) unless $tx->success;
			my $data = WWW::Salesforce::SOAP::response_login($tx->res->dom);
			$self->_instance_url(Mojo::URL->new($data->{serverUrl}));
			$self->_access_token($data->{sessionId});
			$self->_access_time(time); #convert milliseconds to seconds
			return $self->$cb(undef, $self->_access_token);
	}
	)->catch(sub {
		return $self->$cb(pop);
	})->wait;
	return $self;
}


1;

=encoding utf8

=head1 NAME

WWW::Salesforce::Connector - A role to handle some of the implementation of WWW::Salesforce

=head1 SYNOPSIS

	#!/usr/bin/env perl
	use Mojo::Base -strict;
	use WWW::Salesforce;
	use Try::Tiny qw(try catch);

	# via soap
	my $sf_soap = WWW::Salesforce->new(
		login_type => 'soap',
		login_url => Mojo::URL->new('https://login.salesforce.com'),
		version => '34.0',
		username => 'foo@bar.com',
		password => 'mypassword',
		pass_token => 'mypasswordtoken123214123521345',
	);
	# via OAuth2 username and password
	my $sf_oauth2 = WWW::Salesforce->new(
		login_type => 'oauth2_up',
		login_url => Mojo::URL->new('https://login.salesforce.com'),
		version => '34.0',
		consumer_key => 'alksdlkj3hasdg;jlaksghajdhgaghasdg.asdgfasodihgaopih.asdf',
		consumer_secret => 'asdfasdjkfh234123513245',
		username => 'foo@bar.com',
		password => 'mypassword',
		pass_token => 'mypasswordtoken123214123521345',
	);

	# blocking method
	# calling login() will happen automatically.
	try {
		my $res_soap = $sf_soap->query('Select Id, Name, Phone from Account');
		say "found ", scalar(@{$res_soap}), " results via SOAP then RESTful API.";
		my $res_oauth = $sf_oauth2->query('Select Id, Name, Phone from Account');
		say "found ", scalar(@{$res_oauth}), " results via OAuth2 then RESTful API.";
	}
	catch {
		die "Couldn't query the service: $_";
	};

	# non-blocking method
	# calling login() will happen automatically
	Mojo::IOLoop->delay(
		sub {
			my $delay = shift;
			$sf_soap->query('select Id from Account', $delay->begin);
			$sf_oauth2->query('select Id from Account', $delay->begin);
		},
		sub {
			my ($delay, $err,$soap,$err2,$oauth) = @_;
			Carp::croak( $err ) if $err; # make it fatal
			Carp::croak( $err2 ) if $err2; # make it fatal
			say scalar(@$soap), " from soap";
			say scalar(@$oauth), " from oauth2";
		},
	)->catch(sub {say "uh oh, ",pop;})->wait;


=head1 DESCRIPTION

L<WWW::Salesforce::Connector> is a L<Moo::Role> that implements the C<login> and C<logout> methods.

=head1 ATTRIBUTES

L<WWW::Salesforce::Connector> makes the following attributes available.

=head1 METHODS

L<WWW::Salesforce::Connector> makes the following methods available.

=head2 login

	# blocking
	try {
		$sf = $sf->login(); # allows for method-chaining
	} catch {
		die "Errors: $_";
	};

	# non-blocking
	$sf->login(sub {
		my ($sf, $err, $token) = @_;
		say "Our auth token is: $token";
	});

This method will and go through the L<Salesforce Username-Password OAuth Authentication Flow|http://www.salesforce.com/us/developer/docs/api_rest/Content/intro_understanding_username_password_oauth_flow.htm>
process if it needs to.
Calling this method on your own is not necessary as any API call will call C<login> if necessary.  This could be helpful if you're changing C<api_host>s on your instance.
This method will update your C<access_token> on a successful login.

=head2 logout

	# blocking
	try {
		$sf = $sf->logout(); # allows for method chaining.
	} catch {
		die "Errors: $_";
	};

	# non-blocking
	$sf->logout(sub {
		my ( $sf, $err ) = @_;
		say "We're logged out" unless $err;
	});

This method will go through the L<Token Revocation Process|https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_understanding_oauth_endpoints.htm>.
It also removes knowledge of your access token so that you can login again on your next API call.

=head1 AUTHOR

Chase Whitener << <cwhitener at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests on GitHub L<https://github.com/genio/www-salesforce-nb/issues>.
I appreciate any and all criticism, bug reports, enhancements, or fixes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

		perldoc WWW::Salesforce

You can also look for information at:

=over 4

=item * GitHub

L<https://github.com/genio/www-salesforce-nb>

=back

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2015

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<https://github.com/kraih/mojo>, L<Mojolicious::Guides>,
L<http://mojolicio.us>.

=cut
